FISH-788 Support sub-directories for MPCONFIG SecretDirConfigSource. #5006

poikilotherm opened the pull request:
<!--- Title your PR with a Jira reference (if available) followed by brief description - for example: "PAYARA-1234 Add readme file" -->

## Description

Closes #5006 

This adds support for mounting your secrets (or configmapped values) in subdirectories to create scopes.

The relative path is used as the property name. A secret file `SECRET_DIR/foo/bar/secret`
will be available as MPCONFIG value `foo.bar.secret`.

You can still put a file `foo.bar.secret` in `SECRET_DIR` and have the same value available.
Any combination of dots and directories is supported, also the most specific (read: most subdirs and least dots in file and dir names) will win. Updates during runtime are respected and evaluated with longest match, too. An example: `SECRET_DIR/foo/bar.secret` or SECRET_DIR/foo.bar/secret` is fine, too.

## Important Info
### Blockers
None.

## Testing
### New tests
Included.

### Testing Performed
Run tests in `SecretsDirConfigSourceTest`

### Testing Environment
- AdoptOpenJDK (build 11.0.9+11)
- Linux, Fedora 33
- Maven 3.6.3

## Documentation
<!-- Link documentation if a PR exists -->
See payara/Payara-Community-Documentation#114

## Notes for Reviewers
<!-- Any further information for reviewers such as where to start reviewing. Commits should already be clean and the code should already be understandable without this. -->
None.

poikilotherm commented:
IIRC I signed the CLA already when I contributed  to the former https://github.com/payara/docker-payaraserver-full repo.
Although I would love more of that Payara swag :smile_cat:

AlanRoth commented:
Ah, indeed you have. My apologies.

AlanRoth commented:
jenkins test please

AlanRoth commented:
jenkins test please

AlanRoth commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -43,12 +43,10 @@
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
+import java.nio.file.*;
```
Personally, I would avoid using wildcard imports - however, I'll leave it to a consensus

poikilotherm replied:
Oh that was IntelliJ automagic. Happy to switch back again.

poikilotherm replied:
Done with 74288cb

jbee commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -143,39 +153,49 @@ private void findFile() {
-            } catch (IOException ex) {
-                Logger.getLogger(SecretsDirConfigSource.class.getName()).log(Level.SEVERE, null, ex);
+        if (Files.exists(file) && Files.isRegularFile(file) && Files.isReadable(file)) {
+            StringBuilder collector = new StringBuilder();
+            for (String line : Files.readAllLines(file)) {
+                collector.append(line);
             }
+            result = collector.toString();
```
This should use `Files.readAllBytes` to create a new `String`

poikilotherm replied:
OK will look into it. :smiley_cat:

poikilotherm replied:
This has been taken care of.

jbee commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -97,28 +103,32 @@ public String getValue(String property) {
+                // check existence (secret mount might have gone away) and the last modified time
                 FileTime ft = storedModifiedTimes.get(property);
-                Path path = Paths.get(secretsDir.toString(), property);
+                Path path = storedPaths.get(property);
```
to be thread-safe `get` should be avoided and `computeIfAbseent` should take its place.

poikilotherm replied:
"Noice". Didn't knew about `computeIfAbsent`.

poikilotherm replied:
Outdated, as the properties are stored in an inner class now.

jbee reviewed (changes requested):
Thanks @poikilotherm, I think this is a good initiative that needs to mature a little bit more. 

As it is I find it confusing to reason about since the moving parts are checked and manipulated in many places. Many of the flaws I pointed out did not start with your changes but making the essential logic more complex really highlights that this class needs some more attention. 

First of all what consistency do we really want to support? Are files allowed to be updated at runtime? If so what is the observable effect? What about misses and removed entries? And finally collisions. 

As far as I understood you want to handle collisions as first (on time axis) found wins until eventually that very source would no longer hold the property which would open for other files to supply it? What about if two files from the very start contain same property? Which one wins? Do we make sure we process the files in the correct order for this effect?

AlanRoth commented:
jenkins test please

jbee commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -97,28 +103,32 @@ public String getValue(String property) {
+                    // file has been modified since last check, re-read content
+                    result = readFile(path);
                     storedModifiedTimes.put(property, Files.getLastModifiedTime(path));
                     properties.put(property, result);
```
Left this comment before I realised that each file is a single property and forgot to remove it later.

poikilotherm replied:
OK I'll resolve this comment then.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -65,9 +70,10 @@
     private Path secretsDir;
     private ConcurrentHashMap<String, String> properties;
     private ConcurrentHashMap<String, FileTime> storedModifiedTimes;
+    private ConcurrentHashMap<String, Path> storedPaths;
```
Ok, will create a small inner class for this.

Curious: why should this be more thread safe?

jbee replied:
The concurrent map only makes updating entries an atomic operation. So you can update map1 while another thread is reading map1 (updated) and map2 (not yet updated) so the picture this thread sees is only partially updated and generally must be considered inconsistent. In most cases such inconsistencies are exactly what we cannot have and avoiding it is what we mean when we say it is _thread-safe_. If we have one map with an entry with 3 fields updating these 3 is "atomic" and no other thread can observe one field already updated while the other fields have not been updated yet. For the same reason `get` + `put` isn't a good idea as essentially in the meantime the situation you did read might no longer be valid. This disappears when using `compute` or `computeIfAbsent`.

poikilotherm replied:
This has been taken care of.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java:
```diff
@@ -97,28 +103,32 @@ public String getValue(String property) {
             // check whether there is a file there now as there wasn't before
-            Path path = Paths.get(secretsDir.toString(), property);
-            if (Files.exists(path) && Files.isRegularFile(path) && Files.isReadable(path)) {
+            // --> the list of possible paths is used "first match, first serve".
```
I'm not sure what you are trying to do with that information. A value that had no property file on disk before is always a miss in the map cache, returning null, leading to execution of this else branch.

jbee replied:
Yeah it depends on what we want. If we want to make sure we "see" added files as soon as they are written we need to check for it which makes me think that the best way to do update at all is not checking but registering a file watcher that is watching the secret dir directory for changes and updates the map as they occur. No per request checks needed any more reducing the file IO to just the IO actually needed.

poikilotherm replied:
The file watcher has been implemented.

poikilotherm commented:
> Thanks @poikilotherm, I think this is a good initiative that needs to mature a little bit more.

Hi @jbee,
thank you for your extensive review :+1: 

> As it is I find it confusing to reason about since the moving parts are checked and manipulated in many places. Many of the flaws I pointed out did not start with your changes but making the essential logic more complex really highlights that this class needs some more attention.

Absolutely. For starters: should we rename this class to ConfigDirSource to be consistent with docs? In 5.184 the `asadmin` command changed to what it is now, docs weren't updated and this class sticked with the old name. Maybe it's time for a change...

> First of all what consistency do we really want to support? Are files allowed to be updated at runtime? If so what is the observable effect?

Files updated at runtime definitely is a must-have.
See https://kubernetes.io/docs/concepts/configuration/secret/#mounted-secrets-are-updated-automatically

> What about misses and removed entries?

Removed entries definitely is a problem. As the code is working now, removed entries are still cached. On second thought, that might be undesireable in case someone actually tries to "move" a config value (it's not necessarily a secret, also this is the primary use case) to another source with lower ordinal.

> And finally collisions.
> As far as I understood you want to handle collisions as first (on time axis) found wins until eventually that very source would no longer hold the property which would open for other files to supply it? What about if two files from the very start contain same property? Which one wins? Do we make sure we process the files in the correct order for this effect?

Collisions can happen in one scenario only: someone creates a file `CONFIG_DIR/foo.bar.test` and a folder structure with `CONFIG_DIR/foo/bar/test`. This case is handled in https://github.com/payara/Payara/blob/74288cbe744d64f82088a717663c822dcf8462f6/nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/SecretsDirConfigSource.java#L120-L122

First match wins, so someone using the source as available since v5.183 experiences the same behaviour, maintaining backward compatibility. I'd be happy to remove this and allow for the new subdir style only (so no more files with dots in flat file hierarchy).
See also my added doc bits in payara/Payara-Community-Documentation#114

One edge case I didn't look at so far: it might be necessary to add support for ignoring file extensions. Someone might mount a file with a Docker volume as "property.txt", which could not be found. Let me know if you think this is a thing.

There is a small chance for this slippery edge case to make unexpected behaviour happen: someone adding a file to `CONFIG_DIR/foo/bar.test` on application server start would make that file available as `foo.bar.test`, as the file already contains a dot, which is not removed. This cannot happen when added after server start, trying to read the property file on cache miss, as the property name is transformed with all dots replaced to a file path.

poikilotherm commented:
Small addition: we might also want to check that the file we are reading is not bigger than 512KB (which would already be quite huge for text configs) to prevent loading binary files etc.

jbee commented:
@poikilotherm Good points. After getting more familiar with the case I think we should opt for the file watcher to update the map "in the background" while lookups are simply served from the map with no file IO checks. Map has entries with 3 fields. If it is in the map it is returned, if not we have no such property. When file watcher calls back for file changes we check all possible variations of the file for existence (dots or path variants) and always apply the same rule which one wins. This gives us "immediate update" while also always reflecting the same logic for what file wins even in case a new file is added at runtime that wins over a file existing prior to it. Does that make sense?

As a side benefit we do get much clearer code where some methods only read the map and the file watcher code only updates the map (apart from the initial initialisation).

poikilotherm commented:
@jbee I fully agree with you. Will update my PR accordingly.

@jbee @AlanRoth as we need a background runner for the file watcher, is there any preferred way of creating one in Payara? Otherwise I would just go with [`ExecutorService`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html).

poikilotherm commented:
@jbee @AlanRoth I completed a first draft of the rewrite. Please review :smile: 

I didn't yet refactor the tests, please ignore 'em.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+    private ConcurrentHashMap<String, DirProperty> properties = new ConcurrentHashMap<>();
+    private static Logger logger = Logger.getLogger(DirConfigSource.class.getName());
+    
+    public DirConfigSource(PayaraExecutorService executorService) {
+        // get the directory from the app server config
+        findDir();
+        
+        try {
+            // create the watcher for the directory
+            executorService.submit(new DirPropertyWatcher(this.directory));
+            // initial loading
+            initializePropertiesFromPath(this.directory);
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Error during setup of MicroProfile Config Directory Source", e);
+        }
+    }
```
This might as well be moved to an anonymous definition inside `initializePropertiesFromPath()`. Had been moved here for potential reuse, but didn't happen.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+        }
+    }
+    
+    class DirConfigFileVisitor extends SimpleFileVisitor<Path> {
+        /**
+         * Ignore hidden directories
+         */
+        @Override
+        public FileVisitResult preVisitDirectory(java.nio.file.Path dir, BasicFileAttributes attrs) throws IOException {
+            return dir.toFile().isHidden() ? FileVisitResult.SKIP_SUBTREE : FileVisitResult.CONTINUE;
+        }
+    
+        @Override
+        public FileVisitResult visitFile(Path path, BasicFileAttributes mainAtts) throws IOException {
+            updatePropertyFromPath(path, mainAtts);
+            return FileVisitResult.CONTINUE;
+        }
+    }
```
This might as well be moved to an anonymous definition inside initializePropertiesFromPath(). Had been moved here for potential reuse, but didn't happen.

jbee commented:
Doing a review but it will take 1-2 more days before I get it done...

jbee commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,377 @@
+    DirProperty readPropertyFromPath(Path path, BasicFileAttributes mainAtts) throws IOException {
+        if (Files.exists(path) && Files.isRegularFile(path) && Files.isReadable(path)) {
+            return new DirProperty(
+                new String(Files.readAllBytes(path)),
```
I think you want to specify the encoding explicitly here.

poikilotherm replied:
Using UTF-8.

jbee reviewed (commented):
Overall this is what I was thinking about, looks good. Don't be discouraged by the number of detail suggestions I made. Its mostly missing modifiers. I hope you can improve the initialisation of the source a bit. Maybe follow my suggestion to add an "init" method to add the watcher part so you can use the class with or without?

Ping again if you feel we can do a final review including tests.

poikilotherm commented:
Thanks @jbee for your extensive review. Glad you like what you see... Will address next week, so this might make it into 5.2020.7 :smile:

AlanRoth commented:
> so this might make it into 5.2020.7 😄

@poikilotherm, 5.2020.8 definitely 😉 5.2020.7 is ready and releasing very soon. Sorry to be a bearer of bad news, but you can take your time!

poikilotherm commented:
@jbee one more question. There are a lot of occurrences of "secret dir" in class names, functions etc. Should I clean this up in the same go or is this beyond scope?

AlanRoth commented:
If you are able to refactor all occurrences then sure, but I'm not too sure what you mean and what the changes would be?

poikilotherm commented:
> If you are able to refactor all occurrences then sure, but I'm not too sure what you mean and what the changes would be?

Hi @AlanRoth,
some examples:

```
$ rg -i secretdir
nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/admin/GetConfigOrdinal.java
138:                    result = Integer.parseInt(serviceConfig.getSecretDirOrdinality());

nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/MicroprofileConfigConfiguration.java
67:    String getSecretDir();
68:    void setSecretDir(String directory);
71:    String getSecretDirOrdinality();
72:    void setSecretDirOrdinality(String message);

nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/admin/SetConfigOrdinal.java
149:                                config.setSecretDirOrdinality(Integer.toString(ordinal));

nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/admin/GetConfigSecretsDirectory.java
94:            result = serviceConfig.getSecretDir();

nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/admin/SetConfigSecretsDirectory.java
130:                    t.setSecretDir(toSet.toString());

nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java
237:        return Integer.parseInt(configService.getMPConfig().getSecretDirOrdinality());
253:                                    Paths.get(configService.getMPConfig().getSecretDir()),
255:                                    Paths.get(System.getProperty("com.sun.aas.instanceRoot"), configService.getMPConfig().getSecretDir())

appserver/admingui/microprofile-console-plugin-l10n/src/main/resources/fish/payara/admingui/microprofile/Strings_ja.properties
108:config.ordinal.configuration.secretDirectory=\u30b7\u30fc\u30af\u30ec\u30c3\u30c8\u30fb\u30c7\u30a3\u30ec\u30af\u30c8\u30ea

appserver/admingui/microprofile-console-plugin/src/main/resources/fish/payara/admingui/microprofile/Strings.properties
138:config.ordinal.configuration.secretDirectory=Secret Directory

appserver/admingui/microprofile-console-plugin/src/main/resources/microprofile/specs/configAPI/ordinalConfiguration.inc
115:        <sun:property id="secretDirectoryOrdinalProp" labelAlign="left" noWrap="#{true}" overlapLabel="#{false}" 
116:                      label="$resource{i18n_microprofile.config.ordinal.configuration.secretDirectory}"> 
118:                           text="#{pageSession.microprofileConfigMap['secretDirOrdinality']}"/>

```

Propably there's more like it. I didn't touch any of this to make this PR more narrow scoped, but you might prefer otherwise.

AlanRoth commented:
I'm guessing the changes you would like to make is `SecretDir` -> `SecretsDir'? I think it would be better as a second PR linked to this one, just to not obscure changes in this one as there seems to be many occurrences and its a little out of scope. Just don't forget to update and occurrences in this PR (#5007) also!

poikilotherm commented:
@AlanRoth the idea was to remove the "secret" part. Since 5.184 the asadmin command changed to not include it, as the use cases are far more widespread than secrets only. That's why I removed `SecretsDirConfigSource` and switched to `DirConfigSource`.

AlanRoth commented:
Ah, I see, then feel free to make those changes - I think a separate PR will be best.

jbee commented:
>  There are a lot of occurrences of "secret dir" in class names, functions etc. Should I clean this up in the same go or is this beyond scope?

I agree with @AlanRoth that seperate PR is best for the name changes. Also not sure if any of the occurrences other then the source class name should really change. This is more a question of how we want to communicate the feature. While I agree that it is misleading to call is a "secret-something" thing just because it is usually used this is how it has been marketed so far so I would delegate this question to @MattGill98 and @Pandrex247. Functionally there is no doubt this is simply a source that uses one file per property using the file name and path to deduct the property name the file represents. This might have many more usages then just reading secrets but apparently this was the main usage so far.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+ */
+public class DirConfigSource extends PayaraConfigSource implements ConfigSource {
+
+    class DirProperty {
```
Done, but still scope package to allow easier testing.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+        }
+    }
+    
+    class DirPropertyWatcher implements Runnable {
```
Final ok, but remain in package scope and non-static. Package makes testing easier. Static wouldn't allow us to access the update methods or the `DirConfigSource` class attributes.

poikilotherm replied:
Done as described.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+    
+    class DirPropertyWatcher implements Runnable {
+    
+        private Logger logger = Logger.getLogger(DirConfigSource.class.getName());
```
Final OK. Static: can't do that in non-static inner class.

poikilotherm replied:
Done as described.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+    
+        private Logger logger = Logger.getLogger(DirConfigSource.class.getName());
+        WatchService watcher = FileSystems.getDefault().newWatchService();
+        ConcurrentHashMap<WatchKey, Path> keys = new ConcurrentHashMap<>();
```
"watchedFileKeys"

Pandrex247 commented:
Renaming to just DirConfigSource makes sense to me, though obviously comes with the pain of trying to make sure that all occurrences of "secret" are correctly removed - something we've apparently failed to do previously given the current state of our documentation and admin console which is split on this (sometimes just dir, other times secrets-dir).

jbee commented:
@poikilotherm Sorry, I lost track of this PR - is this ready for another review?

poikilotherm commented:
Hi @jbee, I was a bit crammed with other dev work, too. Thx for pinging me.

This is almost ready, but I stumbled over a problem. The file watchers are nice in theory, but have issues.

For example, saving a file in tests will result in multiple events, as the file is created and modified when content is written.
So we might only monitor for "modification".

But I read about another potential issue, that would be much worse: file watchers don't seem to be supported cross-mounts. https://blog.arkey.fr/2019/09/13/watchservice-and-bind-mount

Now I'm not so sure if our approach is a good idea or needs some testing (sample app in K8s Minikube), before we merge this and it's broken...

AlanRoth commented:
Hi @poikilotherm, did you get a chance to test your solution?

poikilotherm commented:
Hi @AlanRoth I was waiting for a response from @pdudits. Should I go ahead and try to create a sample app for a reproducer?

pdudits commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,387 @@
+        
+        @Override
+        public void run() {
+            while (true) {
+                // wait infinitely until we receive an event (or the executor is shutting down)
```
`PayaraExecutorService` tries to terminate gracefully via `awaitTermination` first. This loop therefore delays shutdown of the server by 5 seconds.
Consider scheduling single `take` every few seconds instead, or polling synchronously during get.

poikilotherm replied:
This issue has been taken care of.

pdudits commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/test/java/fish/payara/nucleus/microprofile/config/source/DirConfigSourceTest.java:
```diff
@@ -0,0 +1,239 @@
+    public void testInitializeProperties_IgnoreHidden() throws IOException {
+        // given
+        // none of these should be picked up (hidden file or dir)
+        writeFile(testDirectory, ".hidden.bar.test", "test");
+        writeFile(testDirectory, ".hidden/bar.test", "test");
+        //when
+        source.initializePropertiesFromPath(testDirectory);
+        //then
+        assertEquals(null, source.getValue("hidden.bar.test"));
```
This fails on Windows, because it has different condition for a hidden file. Consider explicit check for dot at start rather than `isHidden`.

poikilotherm replied:
Done. It *should* work now.

pdudits reviewed (changes requested):
Sorry, bit late to the party, and especially didn't look on the past revisions, so sorry if I'll be circling to things that were changed already.

The other use of `WatchService` is negative cache for classloading within `generated` folders. We remove prefix from the cache on first create/modify event in the folder, and polling is done synchronously before read. That use is much simpler yet required few attempts to get right.

> But I read about another potential issue, that would be much worse: file watchers don't seem to be supported cross-mounts. 

That's an important point. And in case of Kubernetes secrets, the files tend to be symbolic links in a `tmpfs` mount. Would file watcher recognize the change in this scenario? 

The blocker for me is build failing on Windows though.

jbee commented:
I see you got this @pdudits - consider it my Christmas gift to you :D

poikilotherm commented:
I finally found the time and patience to dig into a reproducer.
Please find it here: https://github.com/poikilotherm/reproducer-k8s-java-filewatch

I already picked up some enhancement proposals by @pdudits, so this can be re-applied to the config sources code.
Thinking about creating extensive test cases over there, maybe porting them here.

How would you folks feel about integration tests with @testcontainers and introducing JUnit5 in this particular Maven module?

AlanRoth commented:
Hi @poikilotherm, 
this seems to have become a very large PR, perhaps you could create separate PRs with tests using test containers and Junit5? However, I think unit tests should suffice. Once @pdudits requested changes are implemented do you think it would be good to merge? Even if the feature isn't perfect, we just really want to test it out in the wild :)

Thank you,
Alan

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,352 @@
+        }
+    }
+    
+    class DirConfigFileVisitor extends SimpleFileVisitor<Path> {
```
This is outdated, as the visitor has been moved to anonymous definition.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,377 @@
+    
+    private Path directory;
+    private ConcurrentHashMap<String, DirProperty> properties = new ConcurrentHashMap<>();
+    private static Logger logger = Logger.getLogger(DirConfigSource.class.getName());
```
Almost done like this. Kept `directory` non-final to be able to change outside of constructor.

poikilotherm commented:
Well well, things taking long promise to work out well after all. :birthday: 

@jbee, @AlanRoth, @pdudits I would be honored if you folks could take another look on what I crafted. :bug: 

I did not provide any integration tests, unit tests only (as outlined by @AlanRoth). Plus I have absolutely no idea how to create a working snapshot container image to really test if this works in real life. :man_shrugging: 

Once this is done, I would update the docs to fit the codes behavior. :memo:

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,377 @@
+        return "Directory";
+    }
+
+    private void findDir() {
```
I think this is done as you intended. Please feel free to re-open if not.

pdudits reviewed (commented):
Regarding building the image, yes we have too many building options in the tree. To get (all) docker images run:
```
mvn -PBuildDockerImages -pl :docker-images -am clean install
```
This will create image `payara/server-full:5.2020.7-SNAPSHOT` (and few more) that you can push to your local repository and use for tests.

I will also try to test drive this in Kubernetes during next few days, I need to think the use cases through. What are the cases you have in mind? I mean beyond "There will be directory foo, and file bar beneath", how do you imagine actual users of the feature will use it?

Here would be some I have in mind:

1. I deploy multiple applications to a single server. They have separate ConfigMaps and I mount them into a subdirectory of `secrets`. The applications need to have unique prefixes, which is not the first choice with developers, as they tend to follow package name. I will use only single directory level, because ConfigMaps won't let me use `/` in key. I can mount a little bit in under secrets root (I hope and need to test).
1. Credentails are separated from actual application. There's operational settings in a ConfigMap, and credentials in a Secret. I use same kind of deployment with two subdirectories. Secrets must have common prefix, e. g. `secret.` so that they are separated. This works as long as I have full control over property names (e. g. they are not derived from a class name)
1. I want to tune [Microprofile Fault Tolerance](https://github.com/eclipse/microprofile-fault-tolerance/blob/master/spec/src/main/asciidoc/configuration.asciidoc) for specific environment. I loose, because they use `/` as part of the key and neither the previous or the new functionality will help me here.

I believe sooner or later will need to address 2. and 3. and change this config source in more substantial way and probably incompatible way. I'm not asking for that now as that would be bit more involved, but I'd suggest ignoring `*.properties` files for now, while displaying a warning, that those are not supported yet. Probably later we'd implement explicit support for property files, that would be loaded as is, without prefixes. Overriding rules and especially reacting to adding / removing files might get quite complicated, that's why I'd postpone it to separate/later PR.

poikilotherm commented on nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/source/DirConfigSource.java:
```diff
@@ -0,0 +1,448 @@
+                                    Paths.get(path),
+                                    // let's try it relative to server environment root
+                                    Paths.get(System.getProperty("com.sun.aas.instanceRoot"), path)
+                                );
```
Addressed this in dc78c29. Also added unit tests to make sure this fails early.

poikilotherm commented:
> I will also try to test drive this in Kubernetes during next few days, I need to think the use cases through. What are the cases you have in mind? I mean beyond "There will be directory foo, and file bar beneath", how do you imagine actual users of the feature will use it?
> 
> Here would be some I have in mind:
> 
>     1. I deploy multiple applications to a single server. They have separate ConfigMaps and I mount them into a subdirectory of `secrets`. The applications need to have unique prefixes, which is not the first choice with developers, as they tend to follow package name. I will use only single directory level, because ConfigMaps won't let me use `/` in key. I can mount a little bit in under secrets root (I hope and need to test).

Well different applications are likely to have differently named properties, aren't they? If multiple applications share a property "database.username" you'll be in trouble when needing different values/content. This is a no brainer on K8s or Docker, as every application would be in a different app server / container.

Please keep in mind that mounting your `ConfigMap`s and `Secret`s will create files with the "keys" as filenames and value as file content. You  can ensure that different `ConfigMap`s are mounted in different subdirectories of the "config dir" you configure in Payara.

>     2. Credentails are separated from actual application. There's operational settings in a ConfigMap, and credentials in a Secret. I use same kind of deployment with two subdirectories. Secrets must have common prefix, e. g. `secret.` so that they are separated. This works as long as I have full control over property names (e. g. they are not derived from a class name)

I'm not so sure about the need to separate this. On a non-docker installation, you can mix'n'match config and secrets as you see fit. I haven't tested mounting both secrets AND configs in the same directory. K8s is doing really some fancy stuff with symlinks and hidden subdirectories to enable the hot-swapping. This would be a great test scenario!

>     3. I want to tune [Microprofile Fault Tolerance](https://github.com/eclipse/microprofile-fault-tolerance/blob/master/spec/src/main/asciidoc/configuration.asciidoc) for specific environment. I loose, because they use `/` as part of the key and neither the previous or the new functionality will help me here.

True. We would need to think about some ideas how to allow for using a "/" instead of the ".". As the FT thing is rather static and not meant to be used for both app config and ft config, I would agree to use "/" instead of "." when finding a directory "mp-fault-tolerance" or similar within the topmost directory of the source. The dots of the class names are allowed chars for the directory in the current design of this source.

> I believe sooner or later will need to address 2. and 3. and change this config source in more substantial way and probably incompatible way. 

I don't think this is a very complex thing to address. If mixed mounts for (2) is not an option, we might just enable using as many directories for a config source as you might want to have. IMHO the design of the source is capable to handle this. For (3) I already suggested a small workaround (happy to include here if we agree on sth).

> I'm not asking for that now as that would be bit more involved, but I'd suggest ignoring `*.properties` files for now, while displaying a warning, that those are not supported yet. Probably later we'd implement explicit support for property files, that would be loaded as is, without prefixes. Overriding rules and especially reacting to adding / removing files might get quite complicated, that's why I'd postpone it to separate/later PR.

In my opinion, the config source in this very PR should not enable support for property files (following UNIX philosophy). I know that @pdurbin dreams of a single config file for Dataverse, which currently is not possible as is. *BUT* this should live in a different config source. It would be a great addition :star_struck:

pdudits commented:
>  As the FT thing is rather static and not meant to be used for both app config and ft config,

Actually, setting timeouts, concurrency limits or retry counts is one of the things that would greatly benefit from quick feedback from application.

> In my opinion, the config source in this very PR should not enable support for property files

I agree, I only ask for this PR to leave room for property files, so we don't break things later. The only change would be declaring things ending with `.properties` reserved -- adding `&& !path.getFileName().toString().endsWith(".properties")` to `isAptFile`. We already have (too) many config sources, so just like we didn't add "MultiDirConfigSource" now, I'd rather not add "PropertyDirConfigSource" in later release for refining the semantics of config files in a filesystem.

poikilotherm commented:
> We already have (too) many config sources, so just like we didn't add "MultiDirConfigSource" now, I'd rather not add "PropertyDirConfigSource" in later release for refining the semantics of config files in a filesystem.

Something popped up in my head, let me make sure we're on the same boat: that `MultiDirConfigSource` wouldn't be necessary, as we could simply create multiple instances of `DirConfigSource` from the appserver configuration.

pdudits reviewed (approved):
>  It is unlikely to mix the configuration of MP FT with the MP Config things for your app (different APIs)

Since MP FT is configurable with MP Config (that was the part of spec I linked to), it would be reasonable to be able to be able to use same means as for configuring the app. Especially if this would benefit from run-time configuration as i. e. to throttle the application down.

> I see. I believe we should be good guys then and ignore a more compelete list: .yaml, .yml, .json, .properties, .xml. Anything else? Should we make this configurable?

I think we would take it too far with that. Actually, I have different idea how we'd handle properties files once the time comes - it would be single `.properties` file at path level 0 or 1 -- it's already ignored, and safe to ship with a configmap. Thus I'm approving it in current state, as there were no more problems on Windows.

> Something popped up in my head, let me make sure we're on the same boat: that MultiDirConfigSource wouldn't be necessary, as we could simply create multiple instances of DirConfigSource from the appserver configuration.

That would be another solution, currently this is not possible though, but would also be a resonable feature in some next relase.

pdudits commented:
Jenkins test please

poikilotherm commented:
@pdudits I just created the extension for `isAptFile()` to ignore stuff in 401555d. I added the file extensions as an array, non-configurable, but this should be done IMHO.

pdudits commented:
Thanks, @poikilotherm, like I wrote last time, we'll manage without that restrictions to extensions for now (they in the end still could be valid property names, i. e. `parser.validate.xml = false`). Please reset the PR one commit back.

poikilotherm commented:
OK. No problem.

Now that you mention it: a property file named like `parser.validate.xml` would be cut off to `parser.validate` anyway, as I wanted to add the convenience of letting people use a proper file extensions for text files on Windows.

Should I refactor this to only cutoff `.txt`? Any other extensions? I don't need any of this on Linux or Mac...

pdudits commented:
I don't think we should be cutting away any extensions then (oh, I overlooked that part of code!). Windows users will manage without it, and it will be less surprising.

poikilotherm commented:
Completely reverted ignoring selective files in ce30d87.
Remove the file extension cutoff in 45d7a33.

pdudits commented:
Jenkins test please

poikilotherm commented:
So looks like this is getting somewhere.
For whatever reason GitHub didn't let me re-request a review from @jbee 

What's the way forward from here? I assume I should create another PR to the docs repo. Anything else I should try or do?

pdudits commented:
Yes, updating the docs is the way to go next. I'll take care of merging this one in.

AlanRoth commented:
Hi @poikilotherm,

Congratulations on getting the PR through! 96 Comments! Our second most commented PR.

Head over to our documentation, [here](https://github.com/payara/Payara-Community-Documentation) and submit a PR. Let me know if you require any assistance.

Thank you,
Alan

poikilotherm commented:
I love you folks! :tada: 
Thank you very much for your support and pushing me to do my best. Learned a lot from this.

![party](https://user-images.githubusercontent.com/5468333/106119960-14dc5d80-6156-11eb-97db-7eb35b861e7c.gif)

lprimak commented:
Another comment (sorry) - does not pass QA:
Causes this error on default installation with every app deploy:

```
[#|2021-01-28T14:21:49.724-0600|SEVERE|Payara 5.2020.8-SNAPSHOT|fish.payara.nucleus.microprofile.config.source.DirConfigSource|_ThreadID=90;_ThreadName=admin-thread-pool::admin-listener(1);_TimeMillis=1611865309724;_LevelValue=1000;|
  MPCONFIG DirConfigSource: error during setup.
java.io.IOException: Given MPCONFIG directory 'secrets' is no directory, cannot be read or has a leading dot.
	at fish.payara.nucleus.microprofile.config.source.DirConfigSource.findDir(DirConfigSource.java:287)
	at fish.payara.nucleus.microprofile.config.source.DirConfigSource.<init>(DirConfigSource.java:219)
	at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getDefaultSources(ConfigProviderResolverImpl.java:334)
	at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getDefaultSources(ConfigProviderResolverImpl.java:370)
	at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:283)
	at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:297)
	at fish.payara.nucleus.microprofile.config.spi.ConfigProviderResolverImpl.getConfig(ConfigProviderResolverImpl.java:206)
	at org.eclipse.microprofile.config.ConfigProvider.getConfig(ConfigProvider.java:91)
	at fish.payara.microprofile.config.activation.ConfigApplicationContainer.start(ConfigApplicationContainer.java:57)
	at org.glassfish.internal.data.EngineRef.start(EngineRef.java:123)
	at org.glassfish.internal.data.ModuleInfo.start(ModuleInfo.java:293)
	at org.glassfish.internal.data.ApplicationInfo.start(ApplicationInfo.java:364)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.initialize(ApplicationLifecycle.java:623)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:583)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:556)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:552)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:551)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:582)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:574)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:360)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:573)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1497)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:120)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1879)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1755)
	at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:272)
	at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:240)
	at org.glassfish.admin.rest.utils.ResourceUtil.runCommand(ResourceUtil.java:294)
	at org.glassfish.admin.rest.resources.TemplateListOfResource.createResource(TemplateListOfResource.java:332)
	at org.glassfish.admin.rest.resources.TemplateListOfResource.post(TemplateListOfResource.java:166)
	at sun.reflect.GeneratedMethodAccessor323.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356)
	at org.glassfish.admin.rest.adapter.RestAdapter$2.service(RestAdapter.java:335)
	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:189)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
	at java.lang.Thread.run(Thread.java:748)
|#]

```

poikilotherm commented:
Educated guess: forgot that every config source is initialized, no matter if enabled or not. So I have to add a null handler to the constructor or findDir(), to silently ignore this.

What happens on configuration change? Would a user expect the source to start working when configuring via asadmin without a restart of the server?

lprimak commented:
> What happens on configuration change? Would a user expect the source to start working when configuring via asadmin without a restart of the server?

yes absolutely

pdudits commented:
> 
> 
> > What happens on configuration change? Would a user expect the source to start working when configuring via asadmin without a restart of the server?
> 
> yes absolutely

But that would be feature of its own, as entire Config Provider does not update its "meta-configuration" at runtime, rather at app deployment time.

poikilotherm commented:
OK I'll open a new PR, as this one is merged and closed.

I'll only work on the "not-configured" case, leaving the "hot config reload" for later.

pdudits commented:
I do have one in baking actually, need to test it yet.

poikilotherm commented:
> I do have one in baking actually, need to test it yet.

Shall I prepare sth. for supporting it in this source?

poikilotherm commented:
@pdudits just created #5104, plz review :smile:

pdudits commented:
![image](https://user-images.githubusercontent.com/1588543/106270346-28f38e00-622e-11eb-8cbe-ade47f0df1d2.png)

Oh, you beaten me by two minutes! And yours is a bit nicer. But I need to retest it now :)

poikilotherm commented:
I misunderstood that part then. I read you worked on the hot reload stuff, but you worked on the source itself... Sorry 😐
